src/input.rs

```diff
@@ -466,18 +492,49 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
     }
 
     pub fn mouse_input(&mut self, state: ElementState, button: MouseButton, modifiers: ModifiersState) {
-        if let MouseButton::Left = button {
-            let state = mem::replace(&mut self.ctx.mouse_mut().left_button_state, state);
-            if self.ctx.mouse_mut().left_button_state != state {
-                match self.ctx.mouse_mut().left_button_state {
-                    ElementState::Pressed => {
-                        self.on_mouse_press(modifiers);
-                    },
-                    ElementState::Released => {
-                        self.on_mouse_release(modifiers);
+        match button {
+            MouseButton::Left => {
+                let state = mem::replace(&mut self.ctx.mouse_mut().left_button_state, state);
+                if self.ctx.mouse_mut().left_button_state != state {
+                    match self.ctx.mouse_mut().left_button_state {
+                        ElementState::Pressed => {
+                            self.on_mouse_press(button, modifiers);
+                        },
+                        ElementState::Released => {
+                            self.on_mouse_release(button, modifiers);
+                        }
+                    }
+                }
+            }
+            MouseButton::Middle => {
+                let state = mem::replace(&mut self.ctx.mouse_mut().middle_button_state, state);
+                if self.ctx.mouse_mut().middle_button_state != state {
+                    match self.ctx.mouse_mut().middle_button_state {
+                        ElementState::Pressed => {
+                            self.on_mouse_press(button, modifiers);
+                        },
+                        ElementState::Released => {
+                            self.on_mouse_release(button, modifiers);
+                        }
                     }
                 }
             }
+            MouseButton::Right => {
+                let state = mem::replace(&mut self.ctx.mouse_mut().right_button_state, state);
+                if self.ctx.mouse_mut().right_button_state != state {
+                    match self.ctx.mouse_mut().right_button_state {
+                        ElementState::Pressed => {
+                            self.on_mouse_press(button, modifiers);
+                        },
+                        ElementState::Released => {
+                            self.on_mouse_release(button, modifiers);
+                        }
+                    }
```

Comment:
It looks to me like the matches inside the state change check all do the same. Might be cleaner to refactor this to update the state of the button when it changed and then handle the rest for all of them at once.

Just an idea:
```rust
let button_state = match button {
  MouseButton::Left => Some(&mut self.ctx.mouse_mut().left_button_state),
  MouseButton::Middle => Some(&mut self.ctx.mouse_mut().middle_button_state),
  ...
  _ => None
};

if let Some(button_state) = button_state && button_state != state {
  button_state = state;
  match button_state {
    ElementState::Pressed => self.on_mouse_press(button, modifiers);
    ...
  }
}
```

Reply:
I knew there was a lot of duplication there, that way of refactoring just didn't occur to me! I'll get right on that.